contracts/src/L2/L2BridgeController.sol

```diff
@@ -0,0 +1,216 @@
+
+        if (registry.ownerOf(tokenId) != address(this)) {
+            revert NotTokenOwner(tokenId);
+        }
+
+        registry.setSubregistry(tokenId, IRegistry(transferData.subregistry));
+        registry.setResolver(tokenId, transferData.resolver);
+        registry.setTokenObserver(tokenId, ITokenObserver(address(0)));
+        registry.safeTransferFrom(address(this), transferData.owner, tokenId, 1, "");
+
+        bytes memory dnsEncodedName = NameUtils.dnsEncodeEthLabel(transferData.label);
+        emit NameEjectedToL2(dnsEncodedName, tokenId);
+    }
+
+    /**
+     * @dev Override onERC1155Received to handle minting scenarios
+     * When from is address(0), it's a mint operation and we should just return success
+     * Otherwise, delegate to the parent implementation for ejection processing
+     */
+    function onERC1155Received(
+        address /* operator */,
+        address from,
+        uint256 tokenId,
+        uint256 /* amount */,
+        bytes calldata data
+    ) external virtual override onlyRegistry returns (bytes4) {
+        // If from is not address(0), it's not a mint operation - process as ejection
+        if (from != address(0)) {
+            _processEjection(tokenId, data);
+        }
+        
+        return this.onERC1155Received.selector;
+    }
+
+    /**
+     * @dev Default implementation of onRenew that does nothing.
+     * Can be overridden in derived contracts for custom behavior.
+     */
+    function onRenew(uint256 tokenId, uint64 expires, address /*renewedBy*/) external virtual {
+        bridge.sendMessage(BridgeEncoder.encodeRenewal(tokenId, expires));
+    }
+
+    /**
+     * @dev Default implementation of onRelinquish that does nothing.
+     * Can be overridden in derived contracts for custom behavior.
+     */
+    function onRelinquish(uint256 /* tokenId */, address /* relinquishedBy */) external virtual {
+        // Default implementation does nothing
+    }
+
+    /**
+     * @dev Validates 2LD structure and checks if label exists
+     * @param name The DNS-encoded name (must be a 2LD like "example.eth")
+     * @return tokenId The token id of the name
+     */
+    function _findAndValidateLabelStructure(
+        bytes memory name
+    ) internal view returns (uint256 tokenId) {
+        // Read the second label which should be "eth"
+        uint256 labelSize = uint8(name[0]);
+        uint256 tldOffset = 1 + labelSize;
+        uint256 tldSize = uint8(name[tldOffset]);
+        
+        // Verify TLD is "eth"
+        (bytes32 tldHash, ) = NameCoder.readLabel(name, tldOffset);
+        if (tldHash != ETH_TLD_HASH) {
+            revert InvalidTLD(name);
+        }
+        
+        // Verify the name ends after TLD (next byte should be 0)
+        if (name[tldOffset + 1 + tldSize] != 0) {
+            revert InvalidTLD(name);
+        }
+        
+        // Read the 2LD label
+        string memory label = NameUtils.readLabel(name, 0);
```

Comment:
If you move this up to the top, you won't have to calculate the TLD offset separately.